programs/ziti-edge-tunnel/windows-scripts.c

```diff
@@ -0,0 +1,356 @@
+    model_map *hostnames = add_svc_req_data->hostnames;
+    char* dns_ip = add_svc_req_data->dns_ip;
+    if (hostnames == NULL || model_map_size(hostnames) == 0) {
+        ZITI_LOG(DEBUG, "No domains specified to add_nrpt_rules, exiting early");
+        return;
+    }
+    int namespace_template_padding = strlen(namespace_template);
+    hostname_list_t host_names_list = LIST_HEAD_INITIALIZER(host_names_list);
+    int current_size = 0;
+    int rule_size = 0;
+    model_map_iter it = model_map_iterator(hostnames);
+    while(it != NULL) {
+        const char* hostname = model_map_it_key(it);
+        if (current_size > MAX_BUCKET_SIZE || rule_size > MAX_POWERSHELL_SCRIPT_LEN) {
+            chunked_add_nrpt_rules(nrpt_loop, &host_names_list, dns_ip);
+            rule_size = strlen(hostname) + namespace_template_padding;
+            current_size = 0;
+        }
+
+        struct hostname_s *hostname_data = calloc(1, sizeof (struct hostname_s));
+        hostname_data->hostname = strdup(hostname);
+        LIST_INSERT_HEAD(&host_names_list, hostname_data, _next);
+        current_size++;
+        rule_size += strlen(hostname) + namespace_template_padding;
+        it = model_map_it_remove(it);
+    }
+    if (current_size > 0) {
+        chunked_add_nrpt_rules(nrpt_loop, &host_names_list, dns_ip);
+    }
+    add_svc_req_data->dns_ip = NULL;
+    free(hostnames);
+    free(add_svc_req_data);
+}
+
+void add_nrpt_rules(uv_async_t *ar) {
+    ZITI_LOG(VERBOSE, "Add nrpt rules");
+
+    struct add_service_nrpt_req *add_svc_req_data = ar->data;
+    uv_loop_t *nrpt_loop = ar->loop;
+
+    uv_close((uv_handle_t *) ar, (uv_close_cb) free);
+
+    add_nrpt_rules_script(nrpt_loop, add_svc_req_data);
+}
+
+void chunked_remove_nrpt_rules(uv_loop_t *ziti_loop, hostname_list_t *hostnames) {
+    char* script = calloc(MAX_POWERSHELL_SCRIPT_LEN, sizeof(char));
+    size_t buf_len = sprintf(script, "$toRemove = @(\n");
+    size_t copied = buf_len;
+    int domains_size = 0;
+
+    while(!LIST_EMPTY(hostnames)) {
+        struct hostname_s *hostname = LIST_FIRST(hostnames);
+        buf_len = sprintf(script + copied, namespace_template, "\n", hostname);
+        copied += buf_len;
+        domains_size++;
+        LIST_REMOVE(hostname, _next);
+    }
+    buf_len = sprintf(script + copied, "%s\n\n", ")");
+    copied += buf_len;
+
+    buf_len = sprintf(script + copied, "ForEach ($ns in $toRemove){\n");
+    copied += buf_len;
+    buf_len = sprintf(script + copied, "Get-DnsClientNrptRule | where Namespace -eq $ns['n'] | Remove-DnsClientNrptRule -Force -ErrorAction SilentlyContinue\n");
+    copied += buf_len;
+    buf_len = sprintf(script + copied, "}\n");
+    copied += buf_len;
+
+    ZITI_LOG(TRACE, "Removing NRPT script containing %d domains. total script size: %d", domains_size, copied);
+
+    char cmd[MAX_POWERSHELL_COMMAND_LEN];
+    snprintf(cmd, sizeof(cmd),"powershell -Command \"%s\"", script);
```

Comment:
We should check return value of `snprintf` to make sure the string wasn't truncated.